### PR TITLE
Reflection performance tweaks & cleanup

### PR DESCRIPTION
**Pull request by jkotas**
- Delete internal "Unsafe" field getters/setters. They performance was very close (within 10%) of the regular getters/setters.
- Made a few performance tweaks in reflection to make reflection faster overall. The regular field getters/setters are faster than what the unsafe ones used to be with this change.

### CODE REVIEW

**Comment by jkotas** [comment]
Related to discussion in https://github.com/dotnet/corert/pull/6747#discussion_r244563333

**Comment by jkotas** [comment]
@dotnet-bot test Ubuntu x64 Checked CoreFX Tests please
@dotnet-bot test Windows_NT x64 Checked CoreFX Tests please

File: src/System.Private.CoreLib/src/System/Reflection/RtFieldInfo.cs [hunk 1]
```diff
@@ -309,10 +273,9 @@ public override Type FieldType
         {
             get
             {
-                if (m_fieldType == null)
-                    m_fieldType = new Signature(this, m_declaringType).FieldType;
+                return m_fieldType ?? Initialize();
 
-                return m_fieldType;
+                Type Initialize() => m_fieldType = new Signature(this, m_declaringType).FieldType;
```
Comment (stephentoub) [inline src/System.Private.CoreLib/src/System/Reflection/RtFieldInfo.cs]: This doesn't need to be NoInlining (in which case it also currently can't be local)?

File: src/System.Private.CoreLib/src/System/RtType.cs [hunk 2]
```diff
@@ -2379,29 +2381,39 @@ internal override bool CacheEquals(object o)
...
+                    {
+                        IntPtr newgcHandle = new RuntimeTypeHandle(this).GetGCHandle(GCHandleType.WeakTrackResurrection);
+                        IntPtr gcHandle = Interlocked.CompareExchange(ref m_cache, newgcHandle, IntPtr.Zero);
+                        if (gcHandle != IntPtr.Zero)
+                            GCHandle.InternalFree(newgcHandle);
+                    }
 
-                Debug.Assert(cache != null);
-                return cache;
+                    RuntimeTypeCache cache = Unsafe.As<RuntimeTypeCache>(GCHandle.InternalGet(m_cache));
```
Comment (stephentoub) [inline src/System.Private.CoreLib/src/System/RtType.cs]: Is the unsafe casting important for perf in this Initialize method?

**Pull request merged**